### PR TITLE
Change largeCUDATensorTest to largeTensorTest+onlyCUDA; add a buffer to large cuda tensor test

### PR DESCRIPTION
**Pull request by xwang233**
Effectively, `@largeCUDATensorTest` = `@largeTensorTest` + `@onlyCUDA`. 

There was this problem where a user got OOM for a `@largeCUDATensorTest('16GB')` on a 16GB V100. This decorator was checking total memory for a GPU device, however in most cases, we can't allocate all of the memory that a GPU has. So, it would be beneficial that we have a buffer on this `@largeTensorTest` check for CUDA. I added a 10% buffer to it.

Definition of `@largeTensorTest`

https://github.com/pytorch/pytorch/blob/d22dd80128a0d1cfdbfc174a20271719b2f9c1e9/torch/testing/_internal/common_device_type.py#L560-L578

`_has_sufficient_memory`

https://github.com/pytorch/pytorch/blob/d22dd80128a0d1cfdbfc174a20271719b2f9c1e9/torch/testing/_internal/common_device_type.py#L535-L557

`@largeCUDATensorTest`

https://github.com/pytorch/pytorch/blob/d22dd80128a0d1cfdbfc174a20271719b2f9c1e9/torch/testing/_internal/common_device_type.py#L526-L532

### CODE REVIEW

**Comment by xwang233** [comment]
reserved

**Comment by mruberry** [comment]
I like spitting the decorator and letting other device types check their memory, too. I don't like the buffer. I understand the motivation behind the buffer, but I think we need to understand why our test suite fails to make entire GPU memory available and address that underlying issue. I don't think we should continue to work around it. 

The buffer can also be removed and this PR refocused to just splitting the decorator. The memory check should be performed on CPU, too.

**Comment by xwang233** [comment]
@mruberry Thanks for the comments on the decorators. 

Usually, a 16GB V100 has a total memory around 16160 MB instead of 16x1024=16384 MB. I'm not sure if this is the same for all V100 GPUs, but you can check other V100 GPUs you have. 

With the current logic, `torch.cuda.get_device_properties(0).total_memory >= size` will always be false if size is decorated as `16GB`. These 16GB large tensor tests were effectively never executed on 16GB V100. However, recent internal tests using hypervisors somehow made their virtual GPU memory exactly 16384 MB, and that triggered this large tensor test and failed. 

I think it is impossible to allocate the entire GPU memory, because CUDA context and other instances may take a portion upon initializations.

**Comment by mruberry** [comment]
> @mruberry Thanks for the comments on the decorators.
> 
> Usually, a 16GB V100 has a total memory around 16160 MB instead of 16x1024=16384 MB. I'm not sure if this is the same for all V100 GPUs, but you can check other V100 GPUs you have.
> 
> With the current logic, `torch.cuda.get_device_properties(0).total_memory >= size` will always be false if size is decorated as `16GB`. These 16GB large tensor tests were effectively never executed on 16GB V100. However, recent internal tests using hypervisors somehow made their virtual GPU memory exactly 16384 MB, and that triggered this large tensor test and failed.
> 
> I think it is impossible to allocate the entire GPU memory, because CUDA context and other instances may take a portion upon initializations.

I appreciate your analysis and I agree with you. So two things, then:

- Should these tests try to allocate smaller tensors?
- Can the check be improved to, for example, perform a gc and then review the available CUDA memory before deciding whether to run the test or not?

File: test/test_nn.py
Comment (Flamefire) [inline test/test_nn.py:10511]: CuDA is not required here

File: test/test_nn.py
Comment (Flamefire) [inline test/test_nn.py:10493]: CUDA is not required here

File: test/test_nn.py
Comment (Flamefire) [inline test/test_nn.py:10291]: CUDA is not required here

File: test/test_tensor_creation_ops.py
Comment (Flamefire) [inline test/test_tensor_creation_ops.py:760]: CUDA may not be required here

File: test/test_torch.py
Comment (Flamefire) [inline test/test_torch.py:9365]: Is CUDA required here?

**Review by Flamefire — CHANGES_REQUESTED** [review]
Some of the CUDA requirements can be removed. See #46536

Added notes, please double check the "may" ones :+1:

**Comment by xwang233** [comment]
@Flamefire Thanks for checking the onlyCUDA tests. I've removed `onlyCUDA` in these tests. However, those CPU tests take 10 sec ~ 60 sec to run on my machine (dual E5v3 with 48 CPUs), so I marked them with `slowTest`. 

@mruberry 
- I checked these tests. They are mainly testing 64-bit indexing or non-split-able in batch dimension for some operators. So I guess we probably can't blindly reduce the tensor size of these tests. 
- I'm not familiar with the overall effect of python `gc.collect()` on CPU side and CUDA side. Perhaps we can seek help from people with better knowledge?

File: torch/testing/_internal/common_device_type.py
Comment (mruberry) [inline torch/testing/_internal/common_device_type.py:563]: The device can't be acquired from the device_type here because the device_type would be 'cuda' but the actual device might be 'cuda:1'. The decorator needs to take a device_type in the decorator, like 'cuda', but then the check here needs to acquire the (primary) device. See:

https://github.com/pytorch/pytorch/blob/6de619e4a4fd7ce55c3a2f86e666fa9f29026724/torch/testing/_internal/common_device_type.py#L335

File: torch/testing/_internal/common_device_type.py
Comment (mruberry) [inline torch/testing/_internal/common_device_type.py:528]: Instead of multiplying by 1.1, what about checking torch.cuda.memory_allocated to see if there's enough space available?

**Comment by mruberry** [comment]
> * I'm not familiar with the overall effect of python `gc.collect()` on CPU side and CUDA side. Perhaps we can seek help from people with better knowledge?

I made a recommendation for checking the actually available memory and suggested a fix to the decorator, which needs to acquire the device that will be passed to the test.

**Comment by xwang233** [comment]
Seems like `self.get_primary_device()` is not always correct. I got 
```
'TestNN' object has no attribute 'get_primary_device'
```

**Comment by mruberry** [comment]
> Seems like `self.get_primary_device()` is not always correct. I got
> 
> ```
> 'TestNN' object has no attribute 'get_primary_device'
> ```

That seems like you're applying the decorator to a test that's not in a generic test framework class. The classes instantiated by the generic device framework have names like TestNNCPU or TestNNCUDA.

**Comment by xwang233** [comment]
You are right. So I added a optional kwarg `device=` to check the available memory on that device. There won't be a `cuda:0` or `cuda:1` issue, since the test code also uses `cuda` for device?

https://github.com/pytorch/pytorch/blob/eb0c1b65beab38b155ec42db562be466f73dbd81/test/test_nn.py#L3570-L3575

**Comment by mruberry** [comment]
> You are right. So I added a optional kwarg `device=` to check the available memory on that device. There won't be a `cuda:0` or `cuda:1` issue, since the test code also uses `cuda` for device?
> 
> https://github.com/pytorch/pytorch/blob/eb0c1b65beab38b155ec42db562be466f73dbd81/test/test_nn.py#L3570-L3575

You are correct. This PR should document clearly that device is an override for the device generic test framework or an option when outside that framework. That is, by default it will check the memory on the device generic test framework's primary device.

File: torch/testing/_internal/common_device_type.py
Comment (mruberry) [inline torch/testing/_internal/common_device_type.py:527]: `torch.device(device).device_type == 'cuda'`

**Review by mruberry — APPROVED** [review]
Nice work @xwang233! I think we'll have to wait for NVIDIA to add you its CLA, though, before we can merge this. Also this will need a rebase.

**Comment by xwang233** [comment]
Thanks @mruberry ! These test failures seems real 
```
Oct 29 08:25:23 Traceback (most recent call last):
Oct 29 08:25:23   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 274, in instantiated_test
Oct 29 08:25:23     result = test_fn(self, *args)
Oct 29 08:25:23   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 573, in dep_fn
Oct 29 08:25:23     if not _has_sufficient_memory(_device, size_bytes):
Oct 29 08:25:23   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 527, in _has_sufficient_memory
Oct 29 08:25:23     if torch.device(device).device_type == 'cuda':
Oct 29 08:25:23 AttributeError: 'torch.device' object has no attribute 'device_type'
```

**Comment by mruberry** [comment]
> Thanks @mruberry ! These test failures seems real
> 
> ```
> Oct 29 08:25:23 Traceback (most recent call last):
> Oct 29 08:25:23   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 274, in instantiated_test
> Oct 29 08:25:23     result = test_fn(self, *args)
> Oct 29 08:25:23   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 573, in dep_fn
> Oct 29 08:25:23     if not _has_sufficient_memory(_device, size_bytes):
> Oct 29 08:25:23   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 527, in _has_sufficient_memory
> Oct 29 08:25:23     if torch.device(device).device_type == 'cuda':
> Oct 29 08:25:23 AttributeError: 'torch.device' object has no attribute 'device_type'
> ```

They are. It's because my suggestion to use device.device_type was wrong. Sorry about that. It's .type, not .device_type. Should be an easy fix.

**Comment by mruberry** [comment]
@xwang233 @ptrblck Please ping me when the CLA is signed. Unfortunately I can't import this at the moment. If the CLA is already signed also let me know so I can report this as an error.

**Comment by xwang233** [comment]
@mruberry The CLA should be ready.

**Comment by zasdfgbnm** [comment]
Why some tests are marked slow test here? Does slow test skip the test both on CUDA and on CPU? What if a test is only slow on CPU, and is fast on CUDA?

**Comment by mruberry** [comment]
You're correct we don't currently separate slowTest based on device type (if you wanted to add it that'd be fine). We do have a slowtest build that runs in our CI hud. 

We also don't have a great mechanism for determining what's a slowtest and what's not.

**Comment by zasdfgbnm** [comment]
@mruberry What I am complaining about is `test_reduction_split`, `test_conv_transposed_large`, and `test_conv_large`, these are tests designed to test CUDA 32bit indexing bug. CPU does not have this bug, and whether to run this test or not on CPU makes no difference. But this test is marked as slow test because it runs slow on CPU, which really makes no sense. If it is slow on CPU, we should disable it, instead of marking it slow and make it not running by default on CUDA.

**Comment by mruberry** [comment]
That sounds good. You can mark it @onlyCUDA and remove slowtest, then.

**Comment by xwang233** [comment]
@zasdfgbnm Thanks for pointing this out. I agree that add onlyCUDA and remove slowTest is better.

**Comment by ptrblck** [comment]
Some tests started to fail again with e.g.:
```python
RuntimeError: CUDA out of memory. Tried to allocate 16.00 GiB (GPU 0; 16.00 GiB total capacity; 0 bytes already allocated; 13.19 GiB free; 0 bytes reserved in total by PyTorch)
```
as the updated test would not account for the CUDA context (and memory used by other applications) and assumes that PyTorch can allocate the entire device memory.

A potential fix would be to use the returned values from `cudaMemGetInfo` via `torch.cuda.memory.mem_get_info()` to get the free memory reported by the runtime API.

@xwang233 @mruberry do you think this could be a valid approach?

**Comment by xwang233** [comment]
Yeah, I think that could be a good solution. I'm curious why only 13.19 GB out of 16 GB is free? Does that reveal some kind of error in pytorch, or is it just allocated by another program?

**Comment by ngimel** [comment]
Part of it is context size, but almost 3 GB seems excessive (e.g. on my machine where everything is dynamically linked I get
```
RuntimeError: CUDA out of memory. Tried to allocate 18.63 GiB (GPU 0; 15.78 GiB total capacity; 0 bytes already allocated; 14.89 GiB free; 0 bytes reserved in total by PyTorch)
```

**Comment by ptrblck** [comment]
This particular error message was raised in a vGPU setup ([GRID vGPU docs](https://docs.nvidia.com/grid/12.0/grid-vgpu-release-notes-vmware-vsphere/index.html#total-vgpu-frame-buffer-less-than-pgpu-frame-buffer)), which reduces the total available memory as described in the link. 
For this case the loss would be approx.:
```
Max-reserved-fb = vgpu-profile-size-in-mb÷16 + 16 + ecc-adjustments + page-retirement-allocation = (16384/16) +16 + (64 MB overhead for every 1 GB FB i.e 1024) + 128 =   2192
```
and it was just given as an example. Given this and the context size, the `13.19GB` are expected for this setup.
However, even if we ignore vGPU use cases, the current `_has_sufficient_memory` check wouldn't account for e.g. the CUDA context size.
On a non vGPU V100 with 16160MiB the same output as @ngimel.

**Comment by mruberry** [comment]
Updating this check to account for context size (and updating the skip message) sounds like a great idea, @ptrblck

**Pull request closed without merging**
